### PR TITLE
docs(introduction): section on nomenclature

### PR DESCRIPTION
**Pull request by Klaus-Nji-sp**
This is an initial attempt at defining Zeebe nomenclature.  Please do a thorough review.

## Description

An initial section on nomenclature.

## Related issues

Addressing https://github.com/zeebe-io/zeebe/issues/3379

## Pull Request Checklist

- [ ] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [ ] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [ ] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0c3008a949a420566c10c740e10cfdf5e8fddbc7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: docs/src/basics/nomenclature.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+_Workflow Deployment_: a workflow definition cannot execute unless the broker is aware of its existence. This is done by deploying a committing a workflow definition to the broker. The act of pushing of broker is referred to as workflow deployment.
+
+_Broker_:  A broker is an instance of a Zeebe installation which executes workflows and manages workflow state.  A single broker will be installed on a single machine.
+
+_Cluster_:  A cluster represents a configuration of one or more brokers collaborating to execute workflows.  Each broker in a cluster acts as a leader or a follower.  
+
+_Leader_:  In a clustered environment, one broker tis responsible for workflow execution and housekeeping.  Housekeeping includes, taking snapshots, replication and running exports for a given partition.
+
+_Follower_:  In a clustered environment, a broker which is not a leader is a follower of a given partition.  A follower assumes leader responsibilities when the leader goes down.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]:
```suggestion
_Follower_:  In a clustered environment, a broker which is not a leader is a follower of a given partition. A follower can become the new leader when the old leader is not reachable.
```

File: docs/src/basics/nomenclature.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+_Record_: A Zeebe record is a collection of workflow variables and other state transitions of an executing workflow instance at a given point in time.  During execution of a workflow instance numerous records will be generated to capture the various state transitions from start to workflow instance completion.
+
+_Event_:  Within the context of Zeebe, an event represent a state transition associated with an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: We should also describe commands and that events are also reactions to a command.

File: docs/src/basics/nomenclature.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+_Broker_:  A broker is an instance of a Zeebe installation which executes workflows and manages workflow state.  A single broker will be installed on a single machine.
+
+_Cluster_:  A cluster represents a configuration of one or more brokers collaborating to execute workflows.  Each broker in a cluster acts as a leader or a follower.  
+
+_Leader_:  In a clustered environment, one broker tis responsible for workflow execution and housekeeping.  Housekeeping includes, taking snapshots, replication and running exports for a given partition.
+
+_Follower_:  In a clustered environment, a broker which is not a leader is a follower of a given partition.  A follower assumes leader responsibilities when the leader goes down.
+
+_Record_: A Zeebe record is a collection of workflow variables and other state transitions of an executing workflow instance at a given point in time.  During execution of a workflow instance numerous records will be generated to capture the various state transitions from start to workflow instance completion.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]:
```suggestion
_Record_: A record is either a command or an event. For example, a command to create a new workflow instance, or a state transition of an executing workflow instance at a given point in time. During execution of a workflow instance numerous records will be generated to capture the various state transitions from start to workflow instance completion. Records are stored in the log.
```

File: docs/src/basics/nomenclature.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+_Event_:  Within the context of Zeebe, an event represent a state transition associated with an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.
+
+_Log_:  An ordered sequence of events written in persistent storage. This log file is only appended-to and is stored on disk.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]:
```suggestion
_Log_:  An ordered sequence of records written in persistent storage. This log file is only appended-to and is stored on disk. 
```

File: docs/src/basics/nomenclature.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+_Follower_:  In a clustered environment, a broker which is not a leader is a follower of a given partition.  A follower assumes leader responsibilities when the leader goes down.
+
+_Record_: A Zeebe record is a collection of workflow variables and other state transitions of an executing workflow instance at a given point in time.  During execution of a workflow instance numerous records will be generated to capture the various state transitions from start to workflow instance completion.
+
+_Event_:  Within the context of Zeebe, an event represent a state transition associated with an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.
+
+_Log_:  An ordered sequence of events written in persistent storage. This log file is only appended-to and is stored on disk.  
+
+_Segment_:  Data within a log file is segments.  Segments are defined by time period.???
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]:
```suggestion
_Segment_:  The log consists of one or more segments. Each segment is a file that contains records. Segments are deleted when the log is compacted.
```

File: docs/src/basics/nomenclature.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+_Partition_:  A partition represents a logical grouping of data in as Zeebe broker.  This data includes workflow instance variables and events generated by Zeebe.  Partitions are determined by ????
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]:
```suggestion
_Partition_:  A partition represents a logical grouping of data in as Zeebe broker.  This data includes workflow instance variables and events generated by Zeebe. The number of partitions is defined in the configuration.  
```

File: docs/src/basics/nomenclature.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+_Event_:  Within the context of Zeebe, an event represent a state transition associated with an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.
+
+_Log_:  An ordered sequence of events written in persistent storage. This log file is only appended-to and is stored on disk.  
+
+_Segment_:  Data within a log file is segments.  Segments are defined by time period.???
+
+_Partition_:  A partition represents a logical grouping of data in as Zeebe broker.  This data includes workflow instance variables and events generated by Zeebe.  Partitions are determined by ???? 
+
+_Replication_:  In a clustered Zeebe installation configured with more than one broker, data in a partition is copied to the other brokers.  This act of copying data in partitions from one broker to another is also referred to as replication.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Note that the data is replicated from the leader to the followers of a partition.

File: docs/src/basics/nomenclature.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+   
+_Replication Factor_:  This is the number of times data in a partition will be shared and this depends on the number of brokers in a cluster.  The concept of sharing here implies that other brokers should have the exact same copy of data in all partitions.  Data replication allows the system to be resilient to brokers going down. 
+
+_Snapshot_:  The state of all active workflows instances, (these are also known as inflight workflow instances) are stored as records in an in-memory database called  RocksDb.  All events generated as a result of execution of a workflow instance in Zeebe are stored in an append-only log file.  A **snapshot** represents a copy of all data within a partition (including event logs and workflow instance variables in RocksDb) at any given point in time.  Snapshots are binary images stored on disk and can be used to restore execution state of a workflow.   The size of a snapshot is affected by the size of the data.  Size of the data depends on several factors including complexity of the model or business process, the size and quantity of variables in each workflow instance as well as the total number of executing workflow instances in a broker.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Snapshots contain only the in-memory state from RocksDB.

File: docs/src/basics/nomenclature.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+_Exporter_:  An exporter represents a sink to which Zeebe will submitted records representing completed workflow state.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Yes, but not only completed workflow instances. Every record from the log can be exported.

File: docs/src/basics/nomenclature.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+_Exporter_:  An exporter represents a sink to which Zeebe will submitted records representing completed workflow state.  
+
+_Message_:  Within the context of workflow execution, a **message** contains information to be delivered to interested parties.  Messages can be published via Kafka or Zeebe’s internal messaging system.  Messages are associated with timestamp and other constraints such as time-to-live (TTL).
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Yes, in general, messages are ingested by external systems. (e.g. Kafka). There is no internal messaging system ;)

**Review by saig0 — CHANGES_REQUESTED** [review]
Thank you for contributing :+1: 

General notes:
* we need to link the page at `SUMMARY.md`, in order to make it visible, we should add it directly under the root at the bottom
* we should list the terms in alphabetic order to make it easy to find them
* we should add links where you can read more about the topic, if exist

Please have a look at my notes and suggestions :)

File: docs/src/basics/nomenclature.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
+<h1>Nomenclature</h1>
+
+This section defines commonly used terminology referenced within the Zeebe documentation.
+
+**Workflow**:  A workflow is a defined sequence of distinct steps representing your business logic.  Examples of a workflow could be making e-commerce shopping experience, a process to onboard a new employee, etc.  In Zeebe nomenclature, workflows are identified by a unique name and bpmn process id.  The workflow is usually also referred to as the BPMN model.
```
[the remark this answers is not in the document]
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
+<h1>Nomenclature</h1>
```
[the remark this answers is not in the document]
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+This section defines commonly used terminology referenced within the Zeebe documentation.
+
+**Workflow**:  A workflow is a defined sequence of distinct steps representing your business logic.  Examples of a workflow could be making e-commerce shopping experience, a process to onboard a new employee, etc.  In Zeebe nomenclature, workflows are identified by a unique name and bpmn process id.  The workflow is usually also referred to as the BPMN model.
+
+_Workflow Instance_:  While a workflow represents a defined sequence of distinct steps representing your business logic, a workflow instance represents a currently executing or completing workflow.  For a single workflow, there could be many associated workflow instances in various stages of their executing cycle.   Workflow instances are identitied by workflow instance id.  Executing workflows are also sometimes referred to as inflight workflows.
+
+_Workflow Variable_:  A workflow variable represents the execution state of a workflow instance.  The variables capture business process parameters which are input and output of various stages of the workflow instance and which also influence process flow execution.
```
[the remark this answers is not in the document]
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+_Workflow Deployment_: a workflow definition cannot execute unless the broker is aware of its existence. This is done by deploying a committing a workflow definition to the broker. The act of pushing of broker is referred to as workflow deployment.
```
[the remark this answers is not in the document]
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+_Broker_:  A broker is an instance of a Zeebe installation which executes workflows and manages workflow state.  A single broker will be installed on a single machine.
```
[the remark this answers is not in the document]
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Added a note on gateways and clients.  Please review and add more details.

File: docs/src/basics/nomenclature.md (hunk 1 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md (hunk 2 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Please review the event and command section.

File: docs/src/basics/nomenclature.md (hunk 3 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md (hunk 4 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md (hunk 5 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done.

File: docs/src/basics/nomenclature.md (hunk 6 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Done

File: docs/src/basics/nomenclature.md (hunk 7 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Noted.

File: docs/src/basics/nomenclature.md (hunk 8 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Ok.  Please re-review the replication factor.

File: docs/src/basics/nomenclature.md (hunk 9 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: Noted.

File: docs/src/basics/nomenclature.md (hunk 10 above)
Comment (Klaus-Nji-sp) [reply docs/src/basics/nomenclature.md]: I thought one could post a message without Kafka.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0c3008a949a420566c10c740e10cfdf5e8fddbc7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6b447eee6ea0ce0dccd0b16ce1a4826dfcdae9ff
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: docs/src/basics/nomenclature.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+An event represents a state change associated with an aspect of an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.
+
+### Exporter
+An exporter represents a sink to which Zeebe will submitted all records within the log.  This gives users of Zeebe an opportunity to persist records with the log for future use as this data will not be available after log compaction.
+
+### Follower
+In a clustered environment, a broker which is not a leader is a follower of a given partition.  A follower can become the new leader when the old leader is no longer reachable.
+
+### Gateway
+This is the central mechanism through which Zeebe client clients interact with the Zeebe broker.
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Providing more info to gateways.

```suggestion
Clients communicate with the Zeebe cluster through a gateway. The gateway provides a gRPC API and forwards the client commands to the cluster. Depending on the setup, a gateway can be embedded in the broker or standalone.
```

File: docs/src/basics/nomenclature.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+This is the number of times data in a partition will be copied and this depends on the number of brokers in a cluster.  A cluster with one leader and two followers will have a replication factor of three, as data in each partition needs to have three copies.
+
+### Request Timeout
+This is how long a client will wait for a response from the broker after the client has submitted a request.  If a response is not received within the client request timeout, the client can consider the broker unreachable.
+
+### Snapshot
+The state of all active workflows instances, (these are also known as inflight workflow instances) are stored as records in an in-memory database called  RocksDb.  A snapshot represents a copy of all data within the in-memory database at any given point in time.  Snapshots are binary images stored on disk and can be used to restore execution state of a workflow.   The size of a snapshot is affected by the size of the data.  Size of the data depends on several factors including complexity of the model or business process, the size and quantity of variables in each workflow instance as well as the total number of executing workflow instances in a broker.
+
+### Segment
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Change order to 
* Segment
* Snapshot

File: docs/src/basics/nomenclature.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+### Workflow:  
+A workflow is a defined sequence of distinct steps representing your business logic.  Examples of a workflow could be an e-commerce shopping experience, onboarding a new employee, etc.  In Zeebe nomenclature, workflows are identified by a unique name and BPMN process id.  The workflow is usually also referred to as the BPMN model.
+
+### Workflow Instance
+While a workflow represents a defined sequence of distinct steps representing your business logic, a workflow instance represents a currently executing or completed workflow.  For a single workflow, there could be many associated workflow instances in various stages of their executing cycle.   Workflow instances are identitied by workflow instance id.  Executing workflows are also sometimes referred to as inflight workflows.
+
+### Workflow Instance Variable
+A workflow instance variable represents the execution state (i.e data) of a workflow instance.  These variables capture business process parameters which are input and output of various stages of the workflow instance and which also influence process flow execution.
+
+### Worker
```
Comment (saig0) [inline docs/src/basics/nomenclature.md]: Change order to 
* Worker
* Workflow
* Workflow Instance ...

**Review by saig0 — CHANGES_REQUESTED** [review]
@Klaus-Nji-sp, thank you for your fast reaction :+1: 
It looks good. I added some small notes.

Please rename the file to `/docs/src/glossary.md` and link it in `/docs/src/SUMMARY.md`.

Do you want to add some links to the sections?

**Comment by jwulf** [comment]
First of all - can I say: THIS IS AWESOME!!!!!! :-)

You sir, are a legend.

I added some comments.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0c3008a949a420566c10c740e10cfdf5e8fddbc7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6b447eee6ea0ce0dccd0b16ce1a4826dfcdae9ff
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 826f293dbbc028a37be9e595c35a77fed79d0019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0c3008a949a420566c10c740e10cfdf5e8fddbc7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6b447eee6ea0ce0dccd0b16ce1a4826dfcdae9ff
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 826f293dbbc028a37be9e595c35a77fed79d0019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4f009198585a7c9c176f25265155fcb83a37fa07
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0c3008a949a420566c10c740e10cfdf5e8fddbc7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6b447eee6ea0ce0dccd0b16ce1a4826dfcdae9ff
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 826f293dbbc028a37be9e595c35a77fed79d0019
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4f009198585a7c9c176f25265155fcb83a37fa07
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: cb0f4f444f573869449a8e54f4cdc358dc156760
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by Klaus-Nji-sp** [comment]
> First of all - can I say: THIS IS AWESOME!!!!!! :-)
> 
> You sir, are a legend.
> 
> I added some comments.

@jwulf you guys are the legends lol. Just trying to help this awesome effort.  Cannot find your comments BTW.

**Review by saig0 — APPROVED** [review]
Looks good for me :+1:

@jwulf, what do you think?

**Comment by saig0** [comment]
I'm merging the PR now. 

@jwulf feel free to open an issue or a pull request to improve the glossary :)

**Pull request merged**